core/src/main/java/io/grpc/internal/AbstractStream2.java

```diff
@@ -0,0 +1,261 @@
+   * to the remote endpoint.
+   *
+   * @param numBytes the number of bytes being sent.
+   */
+  protected final void onSendingBytes(int numBytes) {
+    transportState().onSendingBytes(numBytes);
+  }
+
+  /**
+   * Only called from transport thread (except for private interactions with AbstractStream2).
+   */
+  public abstract static class TransportState implements MessageDeframer.Listener {
+    /**
+     * The default number of queued bytes for a given stream, below which
+     * {@link StreamListener#onReady()} will be called.
+     */
+    private static final int DEFAULT_ONREADY_THRESHOLD = 32 * 1024;
+
+    private final MessageDeframer deframer;
+    private final Object onReadyLock = new Object();
+    /**
+     * The number of bytes currently queued, waiting to be sent. When this falls below
+     * DEFAULT_ONREADY_THRESHOLD, {@link StreamListener#onReady()} will be called.
+     */
+    @GuardedBy("onReadyLock")
+    private int numSentBytesQueued;
+    /**
+     * Indicates the stream has been created on the connection. This implies that the stream is no
+     * longer limited by MAX_CONCURRENT_STREAMS.
+     */
+    @GuardedBy("onReadyLock")
+    private boolean allocated;
+
+    protected TransportState(int maxMessageSize) {
+      deframer = new MessageDeframer(this, Codec.Identity.NONE, maxMessageSize);
+    }
+
+    @VisibleForTesting
+    TransportState(MessageDeframer deframer) {
+      this.deframer = deframer;
+    }
+
+    /**
+     * Override this method to provide a stream listener.
+     */
+    protected abstract StreamListener listener();
+
+    @Override
+    public void messageRead(InputStream is) {
+      listener().messageRead(is);
+    }
+
+    /**
+     * Called when a {@link #deframe(ReadableBuffer, boolean)} operation failed.
+     *
+     * @param cause the actual failure
+     */
+    protected abstract void deframeFailed(Throwable cause);
+
+    /**
+     * Closes this deframer and frees any resources. After this method is called, additional calls
+     * will have no effect.
+     */
+    protected final void closeDeframer() {
+      deframer.close();
+    }
+
+    /**
+     * Called to parse a received frame and attempt delivery of any completed
+     * messages. Must be called from the transport thread.
+     */
+    protected final void deframe(ReadableBuffer frame, boolean endOfStream) {
+      try {
+        deframer.deframe(frame, endOfStream);
+      } catch (Throwable t) {
+        deframeFailed(t);
+      }
+    }
+
+    /**
+     * Called to request the given number of messages from the deframer. Must be called
+     * from the transport thread.
+     */
+    public final void requestMessagesFromDeframer(int numMessages) {
+      try {
+        deframer.request(numMessages);
+      } catch (Throwable t) {
+        deframeFailed(t);
+      }
+    }
+
+    private final void setDecompressor(Decompressor decompressor) {
```

Comment:
The finals are trying to be an aid that "if you're overriding this, you're probably doing it wrong". It's mainly for readability for you to know that subclasses don't deal with this stuff.

I should actually put it on a few more methods (like `messageRead` above). Although `AbstractServerStream` is abstract, it does end up implementing all methods of `ServerStream`, with no intention of them being overridden.